abad.go

```diff
@@ -97,4 +121,103 @@ func (a *Abad) evalUnaryExpr(expr *ast.UnaryExpr) (Obj, error) {
 	}
 
 	return num, nil
+}
+
+func (a *Abad) evalExpr(n ast.Node) (types.Value, error) {
+	if !ast.IsExpr(n) {
+		panic("internal error: not an expression")
+	}
+
+	switch n.Type() {
+	case ast.NodeNumber:
+		val := n.(ast.Number)
+		return types.Number(val.Value()), nil
+	case ast.NodeIdent:
+		val := n.(ast.Ident)
+		return a.evalIdentExpr(val)
+	case ast.NodeMemberExpr:
+		val := n.(*ast.MemberExpr)
+		return a.evalMemberExpr(val)
+	case ast.NodeCallExpr:
+		val := n.(*ast.CallExpr)
+		return a.evalCallExpr(val)
+	case ast.NodeUnaryExpr:
+		expr := n.(*ast.UnaryExpr)
+		return a.evalUnaryExpr(expr)
+	}
+
+	panic("unreachable")
+	return nil, nil
+}
+
+func (a *Abad) evalIdentExpr(ident ast.Ident) (types.Value, error) {
+	val, err := a.global.Get(utf16.Str(ident))
+	if err != nil {
+		return nil, err
+	}
+
+	if types.StrictEqual(val, types.Undefined) {
+		return nil, fmt.Errorf("%s is not defined",
+			ident.String())
+	}
+
+	return val, nil
+}
+
+func (a *Abad) evalMemberExpr(member *ast.MemberExpr) (types.Value, error) {
+	objval, err := a.evalExpr(member.Object)
+	if err != nil {
+		return nil, err
+	}
+
+	if objval.Kind() != types.KindObject {
+		panic("wrapping primitive values not implemented yet")
+	}
+
+	obj, err := objval.ToObject()
+	if err != nil {
+		return nil, err
+	}
+
+	return obj.Get(utf16.Str(member.Property))
+}
+
+func (a *Abad) evalCallExpr(call *ast.CallExpr) (types.Value, error) {
+	// TODO(i4k): safe to assume the AST is ok?
```

Comment:
If the method is private and you can check clearly on the code that all the paths to it are validated I think you can =). Specially if you have tests to get possible regressions (if someone changes the code). Otherwise our code will be littered with redundant checks all the time =/

Reply:
yes, I usually think this way. But in case of call in  js it's a little annoying because a Function is an Object with a Code property inside (and a Call method). The Code property comes from user declaration of `function xxx() {}`. 
But I agree we can ensure this in the interpreter. It was just a reminder to revisit the path from user code til here.